app/src/main/java/com/duckduckgo/app/browser/downloader/DownloaderUtil.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2022 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.downloader
+
+import android.net.Uri
+import android.webkit.MimeTypeMap
+import timber.log.Timber
+import java.lang.IllegalStateException
+import java.util.*
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+object DownloaderUtil {
```

Comment:
Contains the new approach for extracting the file name. It is largely based on `UrlUtil.guessFileName` but adapted to address the issues we discovered. See https://app.asana.com/0/0/1201532076539209/f